pallets/stablewap/src/lib.rs

```diff
@@ -0,0 +1,705 @@
+		InsufficientShares,
+
+		/// Liquidity has not reached the required minimum.
+		InsufficientLiquidity,
+
+		/// Insufficient liquidity left in the pool after withdrawal.
+		InsufficientLiquidityRemaining,
+
+		/// Amount is less than min trading limit.
+		InsufficientTradingAmount,
+
+		/// Minimum limit has not been reached during trade.
+		BuyLimitNotReached,
+
+		/// Maximum limit has been exceeded during trade.
+		SellLimitExceeded,
+
+		/// Initial liquidity of asset must be > 0.
+		InvalidInitialLiquidity,
+
+		/// Account balance is too low.
+		BalanceTooLow,
+
+		/// Amplification is outside specified range.
+		InvalidAmplification,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create a 2-asset pool with initial liquidity for both assets.
+		///
+		/// Both assets must be correctly registered in `T::AssetRegistry`
+		///
+		/// Initial liquidity must be > 0.
+		///
+		/// Origin is given corresponding amount of shares.
+		///
+		/// Parameters:
+		/// - `origin`: Must be T::CreatePoolOrigin
+		/// - `assets`: Asset ids tuple
+		/// - `initial_liquidity`: Corresponding initial liquidity of `assets`
+		/// - `amplification`: Pool amplification
+		/// - `fee`: trade fee to be applied in sell/buy trades
+		///
+		/// Emits `PoolCreated` event when successful.
+		#[pallet::weight(<T as Config>::WeightInfo::create_pool())]
+		#[transactional]
+		pub fn create_pool(
+			origin: OriginFor<T>,
+			assets: (T::AssetId, T::AssetId),
+			initial_liquidity: (Balance, Balance),
+			amplification: u32,
+			fee: Permill,
+		) -> DispatchResult {
+			let who = T::CreatePoolOrigin::ensure_origin(origin)?;
+
+			let pool_assets: PoolAssets<T::AssetId> = assets.into();
+
+			ensure!(pool_assets.is_valid(), Error::<T>::SameAssets);
+
+			ensure!(
+				T::AmplificationRange::get().contains(&amplification),
+				Error::<T>::InvalidAmplification
+			);
+
+			for asset in (&pool_assets).into_iter() {
+				ensure!(T::AssetRegistry::exists(asset), Error::<T>::AssetNotRegistered);
+			}
+
+			ensure!(
+				initial_liquidity.0 > Balance::zero(),
+				Error::<T>::InvalidInitialLiquidity
+			);
+			ensure!(
+				initial_liquidity.1 > Balance::zero(),
+				Error::<T>::InvalidInitialLiquidity
+			);
+
+			ensure!(
+				T::Currency::free_balance(assets.0, &who) >= initial_liquidity.0,
+				Error::<T>::BalanceTooLow,
+			);
+			ensure!(
+				T::Currency::free_balance(assets.1, &who) >= initial_liquidity.1,
+				Error::<T>::BalanceTooLow,
+			);
+
+			let share_asset_ident = T::ShareAccountId::name(&pool_assets, Some(POOL_IDENTIFIER));
+
+			let share_asset = T::AssetRegistry::get_or_create_shared_asset(
+				share_asset_ident,
+				(&pool_assets).into(),
+				Balance::zero(),
+			)?;
+
+			let pool_id = PoolId(share_asset);
+
+			let pool = Pools::<T>::try_mutate(&pool_id, |maybe_pool| -> Result<PoolInfo<T::AssetId>, DispatchError> {
+				ensure!(maybe_pool.is_none(), Error::<T>::PoolExists);
+
+				let pool = PoolInfo {
+					assets: pool_assets.clone(),
```

Comment:
Do we need to store assets here since we always use poolId which means we know the assets?

Reply:
PoolId is asset id of share token. But you dont really know what assets are in that pool just from the pool id.

What is your idea? how we could determine the two assets ?

Reply:
Ok just noticed we have shareTokenRegistry will need to dive deeper

Reply:
yes we do. but stableswap pallet does not really depend on specific registry implementation. 

we would have to add interface to return assets for given share token. Otherwise registry is not really required to have this information. 

Our registry currently does, as it is stored share token type with the two assets provided.

however, when we we add more assets into stableswap pool - this wont be possible anymore ( registry will have to be adjusted to support multi asset share token in such case).

Reply:
So if I have a share token of aUSD and DAI I can go to asset registry look up that it's stable type then go to poolId, get id of the pool from that and then pull everything else right?

Reply:
Registry will tell you only that the asset id is share asset type for two other assets. But it does not tell you which amm it is. 

well, perhaps if you look at the identifier in the mapping between names and ids - you can figure it because each amm uses its identifier ( xyk, sts).

if you have share token, you already know the pool id in the stableswap.

Reply:
Asking because I was doing very brief skim to get implementation guide for FE and it's not as easy as I thought. Will need to dive deeper. But basically share token - sts. Poll for poolInfo by sharetoken in stableswap. Gold.